### PR TITLE
support cohere models in build_index()

### PR DESCRIPTION
**Pull request by jingyizhu99**
# Description

Please add an informative description that covers that changes made by the pull request and link all relevant issues.

# All Promptflow Contribution checklist:
- [x] **The pull request does not introduce [breaking changes].**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [x] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow-rag/promptflow/rag/_build_mlindex.py [hunk 1]
```diff
@@ -72,15 +72,24 @@ def build_index(
         )
         raise e
 
-    if not embeddings_model_config.embeddings_model:
-        raise ValueError("Please specify embeddings_model_config.embeddings_model")
-    embeddings_model = build_open_ai_protocol(embeddings_model_config.embeddings_model)
+    if not embeddings_model_config.model_name:
+        raise ValueError("Please specify embeddings_model_config.model_name")
+
+    if "cohere" in embeddings_model_config.model_name:
```
Comment (liangjason87) [inline src/promptflow-rag/promptflow/rag/_build_mlindex.py:78]: Is the plan ultimately to have a "supported_models" list that we can easily add new models to?

File: src/promptflow-rag/promptflow/rag/_build_mlindex.py [hunk 2]
```diff
@@ -232,16 +281,29 @@ def _create_mlindex_from_existing_ai_search(
...
-        model_connection_args = {
-            "key": openai.api_key,
+    if "cohere" in embedding_model:
+        # api_key = "SERVERLESS_CONNECTION_KEY"
+        # api_base = "SERVERLESS_CONNECTION_ENDPOINT"
+        # connection_args = {
+        #     "connection_type": "environment",
+        #     "connection": {"key": api_key},
+        #     "endpoint": os.getenv(api_base),
+        # }
```
Comment (liangjason87) [inline src/promptflow-rag/promptflow/rag/_build_mlindex.py:291]: Comment?

File: src/promptflow-rag/promptflow/rag/_build_mlindex.py (hunk 1 above)
Comment (jingyizhu99) [reply src/promptflow-rag/promptflow/rag/_build_mlindex.py:78]: yes, this should be addressed as a follow up task

File: src/promptflow-rag/promptflow/rag/_build_mlindex.py (hunk 2 above)
Comment (jingyizhu99) [reply src/promptflow-rag/promptflow/rag/_build_mlindex.py:291]: this part of the code should be uncommented once azureml-rag support connecting to serverless endpoints with environment variables.

**Review by liangjason87 — APPROVED** [review]

**Review by liangjason87 — APPROVED** [review]

File: src/promptflow-rag/build_index_sample.ipynb [hunk 3]
```diff
@@ -0,0 +1,149 @@
+{
```
[the remark this answers is not in the document]
Comment (jingyizhu99) [reply src/promptflow-rag/build_index_sample.ipynb:1]: yes, I will move them into examples

File: src/promptflow-rag/promptflow/rag/_build_mlindex.py [hunk 4]
```diff
@@ -72,15 +72,27 @@ def build_index(
         )
         raise e
 
-    if not embeddings_model_config.embeddings_model:
```
[the remark this answers is not in the document]
Comment (jingyizhu99) [reply src/promptflow-rag/promptflow/rag/_build_mlindex.py:75]: yes, we still check if this is not null. This is a naming change from embeddings_model to model_name

File: src/promptflow-rag/promptflow/rag/_build_mlindex.py [hunk 5]
```diff
@@ -128,10 +147,27 @@ def build_index(
...
-    embedder = EmbeddingsContainer.from_uri(
-        embeddings_model,
-        **connection_args,
-    )
+        embedder = EmbeddingsContainer.from_uri(
+            embeddings_model_uri,
+            **connection_args,
+        )
+    elif not embeddings_model_uri:
+        # cohere connection doesn't support environment variables yet
```
[the remark this answers is not in the document]
Comment (jingyizhu99) [reply src/promptflow-rag/promptflow/rag/_build_mlindex.py:155]: yes, line 78 we check if "cohere" in embeddings_model_config.model_name and set embeddings_model_uri to None if so. All the serverless connections doesn't need model_uri so the null check here will be enough.

File: src/promptflow-rag/promptflow/rag/_build_mlindex.py [hunk 6]
```diff
@@ -105,18 +117,25 @@ def build_index(
     )
 
     connection_args = {}
-    if "open_ai" in embeddings_model:
-        import os
-
-        if embeddings_model_config.aoai_connection_id:
-            aoai_connection = get_connection_by_id_v2(embeddings_model_config.aoai_connection_id)
+    if embeddings_model_uri and "open_ai" in embeddings_model_uri:
```
[the remark this answers is not in the document]
Comment (jingyizhu99) [reply src/promptflow-rag/promptflow/rag/_build_mlindex.py:120]: I actually don't see improvement in readability, so probably won't change the branching logic here.

**Review by Adrian999999999 — APPROVED** [review]

**Review by Jackxu9946 — APPROVED** [review]

**Review by liangjason87 — DISMISSED** [review]

**Review by liangjason87 — APPROVED** [review]

**Review by Jackxu9946 — APPROVED** [review]

**Review by hagh — APPROVED** [review]

**Review by wangchao1230 — APPROVED** [review]

**Pull request merged**